xbmc/platform/darwin/ios-common/network/NetworkIOS.mm

```diff
@@ -0,0 +1,401 @@
+      struct sockaddr* sa_tab[RTAX_MAX];
+      for (int i = 0; i < RTAX_MAX; i++)
+      {
+        if (rt->rtm_addrs & (1 << i))
+        {
+          sa_tab[i] = sa;
+          sa = reinterpret_cast<struct sockaddr*>(
+              reinterpret_cast<char*>(sa) +
+              ((sa->sa_len) > 0 ? (1 + (((sa->sa_len) - 1) | (sizeof(long) - 1))) : sizeof(long)));
+        }
+        else
+        {
+          sa_tab[i] = nullptr;
+        }
+      }
+
+      if (((rt->rtm_addrs & (RTA_DST | RTA_GATEWAY)) == (RTA_DST | RTA_GATEWAY)) &&
+          sa_tab[RTAX_DST]->sa_family == afinet_type[ip_type] &&
+          sa_tab[RTAX_GATEWAY]->sa_family == afinet_type[ip_type])
+      {
+        if (afinet_type[ip_type] == AF_INET)
+        {
+          if ((reinterpret_cast<struct sockaddr_in*>(sa_tab[RTAX_DST]))->sin_addr.s_addr == 0)
+          {
+            char dstStr4[INET_ADDRSTRLEN];
+            char srcStr4[INET_ADDRSTRLEN];
+            memcpy(srcStr4,
+                   &(reinterpret_cast<struct sockaddr_in*>(sa_tab[RTAX_GATEWAY]))->sin_addr,
+                   sizeof(struct in_addr));
+            if (inet_ntop(AF_INET, srcStr4, dstStr4, INET_ADDRSTRLEN) != nullptr)
+              gateway = dstStr4;
+            break;
+          }
+        }
+        else if (afinet_type[ip_type] == AF_INET6)
+        {
+          if ((reinterpret_cast<struct sockaddr_in*>(sa_tab[RTAX_DST]))->sin_addr.s_addr == 0)
+          {
+            char dstStr6[INET6_ADDRSTRLEN];
+            char srcStr6[INET6_ADDRSTRLEN];
+            memcpy(srcStr6,
+                   &(reinterpret_cast<struct sockaddr_in6*>(sa_tab[RTAX_GATEWAY]))->sin6_addr,
+                   sizeof(struct in6_addr));
+            if (inet_ntop(AF_INET6, srcStr6, dstStr6, INET6_ADDRSTRLEN) != nullptr)
+              gateway = dstStr6;
+            break;
+          }
+        }
+      }
+    }
+    free(buf);
+  }
+
+  return gateway;
+}
+
+
+CNetworkIOS::CNetworkIOS()
+  : CNetworkBase()
+{
+  m_sock = socket(AF_INET, SOCK_DGRAM, 0);
+  queryInterfaceList();
+}
+
+CNetworkIOS::~CNetworkIOS()
+{
+  if (m_sock != -1)
+    close(CNetworkIOS::m_sock);
+
+  for (std::vector<CNetworkInterface*>::iterator it = m_interfaces.begin();
+       it != m_interfaces.end();)
+  {
+    CNetworkInterface* nInt = *it;
+    delete nInt;
+    it = m_interfaces.erase(it);
+  }
+}
+
+std::vector<CNetworkInterface*>& CNetworkIOS::GetInterfaceList()
+{
+  return m_interfaces;
+}
+
+CNetworkInterface* CNetworkIOS::GetFirstConnectedInterface()
+{
+  std::vector<CNetworkInterface*>& ifaces = GetInterfaceList();
+  std::vector<CNetworkInterfaceIOS*>& iosifaces =
+      reinterpret_cast<std::vector<CNetworkInterfaceIOS*>&>(ifaces);
+
+  CNetworkInterfaceIOS* ifWifi = nullptr;
+  CNetworkInterfaceIOS* ifCell = nullptr;
+  CNetworkInterfaceIOS* ifVPN = nullptr;
+
+  for (std::vector<CNetworkInterfaceIOS*>::const_iterator iter = iosifaces.begin();
+       iter != iosifaces.end();)
```

Comment:
use range based for loop
With that you also can't forget to jump to the next element after each iteration.